core/src/main/java/tc/oc/pgm/listeners/SkinCache.java

```diff
@@ -0,0 +1,146 @@
+package tc.oc.pgm.listeners;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.collect.Maps;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import org.bukkit.Bukkit;
+import org.bukkit.Skin;
+import org.bukkit.entity.Player;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.EventPriority;
+import org.bukkit.event.Listener;
+import org.bukkit.event.player.PlayerJoinEvent;
+import org.bukkit.event.player.PlayerQuitEvent;
+import tc.oc.pgm.api.PGM;
+import tc.oc.pgm.api.Permissions;
+import tc.oc.pgm.api.integration.Integration;
+import tc.oc.pgm.api.player.MatchPlayer;
+import tc.oc.pgm.util.event.player.PlayerSkinChangeEvent;
+
+public class SkinCache implements Listener {
+
+  private final Cache<UUID, Skin> offlineSkins =
+      CacheBuilder.newBuilder().maximumSize(300).expireAfterWrite(3, TimeUnit.HOURS).build();
+  private final Random random = new Random();
+
+  private final Map<UUID, Skin> customSkins = Maps.newHashMap();
+
+  // TODO: NEEDS WORK! Backup skins when 0 are online, prevent duplicates, etc
+  private Skin getRandomSkin() {
+    if (offlineSkins.size() == 0) {
+      return Skin.EMPTY; // TODO: Warning, this may be bad for 1.16 clients...
+    }
+    List<Skin> skins = offlineSkins.asMap().values().stream().collect(Collectors.toList());
+    return skins.get(random.nextInt(skins.size()));
+  }
+
+  private Skin getSkin(Player player) {
+    if (customSkins.containsKey(player.getUniqueId())) {
+      return customSkins.get(player.getUniqueId());
+    }
+    return getRandomSkin();
+  }
+
+  private boolean canUseSkin(Player player) {
+    return !player.hasPermission(Permissions.STAFF)
+        && !player.hasPermission(Permissions.PREMIUM); // TODO: add specific node too
+  }
+
+  @EventHandler
+  public void onPlayerQuit(PlayerQuitEvent event) {
+    Player player = event.getPlayer();
+    if (canUseSkin(player)) {
+      offlineSkins.put(player.getUniqueId(), player.getSkin());
+    }
+  }
+
+  @EventHandler(priority = EventPriority.LOW)
+  public void onPlayerJoin(PlayerJoinEvent event) {
+    offlineSkins.invalidate(event.getPlayer().getUniqueId());
+  }
+
+  @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
+  public void refreshNamesOnLogin(PlayerJoinEvent event) {
+    refreshPlayer(event.getPlayer());
+  }
+
+  // SPORTPAPER STUFF - TODO: Add alternative method and check if server is running SportPaper to
```

Comment:
What happens on Spigot? Let's make sure it will safely fail.

Reply:
on spigot this functionality will simply not be supported as spigot doesn't have per-viewer fake display names and skins

If the methods don't exist this should be all just skipped